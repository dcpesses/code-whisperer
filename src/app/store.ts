import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import counterReducer from '@/features/redux-counter/counterSlice';
import modalCommandListReducer from '@/features/modal-command-list/modalSlice';
// import twitchReducer from '../features/twitch/twitchSlice';
import channelReducer from '@/features/twitch/channel-slice.js';
import userReducer from '@/features/player-queue/user-slice.js';
import queueReducer from '@/features/player-queue/queue-slice';
import todosReducer from '@/features/todos/todosSlice';

const reducer = {
  counter: counterReducer,
  channel: channelReducer,
  modal: modalCommandListReducer,
  queue: queueReducer,
  user: userReducer,
  todos: todosReducer,
};
export const store = configureStore({ reducer });

export function getStoreWithState(preloadedState?: RootState) {
  return configureStore({ reducer, preloadedState });
}

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
