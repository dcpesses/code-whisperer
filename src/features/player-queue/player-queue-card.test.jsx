/* eslint-env jest */
import {vi} from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';

import { getStoreWithState } from '@/app/store';
import { Provider } from 'react-redux';
import PlayerQueueCard from './player-queue-card';


describe('PlayerQueueCard', () => {
  let store;
  beforeEach(() => {
    store = getStoreWithState();
  });
  test('Should render without error', () => {
    const {container} = render(
      <Provider store={store}>
        <PlayerQueueCard />
      </Provider>
    );
    expect(container).toMatchSnapshot();
    // let button = screen.getByText('View More');
    fireEvent.click(screen.getByText('View More'));
    expect(container).toMatchSnapshot();
    // let button = screen.getByText('View Less');
    fireEvent.click(screen.getByText('View Less'));
    expect(container).toMatchSnapshot();
  });

  test('Should render as a user in the interested queue', () => {
    store = getStoreWithState({
      user: {
        chatters: {
          dcpesses: {
            id: '473294395',
            login: 'dcpesses',
            display_name: 'dcpesses',
            type: '',
            broadcaster_type: '',
            description: 'Don\'t mind me, I\'m just here for the Jackbox games and to support my peeps. ',
            profile_image_url: 'profile_image-300x300.png',
            offline_image_url: '',
            view_count: 0,
            created_at: '2019-11-18T10:47:34Z'
          }
        },
        whisperStatus: {}
      }
    });
    const actionHandler = vi.fn();
    const removeHandler = vi.fn();
    const {container} = render(
      <Provider store={store}>
        <PlayerQueueCard
          btnProps={{
            label: 'Add 2 Playing',
            onClick: actionHandler,
          }}
          onRemoveUser={removeHandler}
          onSendCode={null}
          queueName={'interested'}
          prioritySeat={false}
          relativeTime={null}
          showSendButton={false}
          username={'dcpesses'}
        />
      </Provider>
    );
    expect(container).toMatchSnapshot();

    const button1 = screen.getByTitle('Remove');
    fireEvent.click(button1);
    expect(removeHandler).toHaveBeenCalled();
    const button2 = screen.getByTitle('Add 2 Playing');
    fireEvent.click(button2);
    expect(actionHandler).toHaveBeenCalled();
  });

  test('Should render as a whispered user in the playing queue', () => {
    store = getStoreWithState({
      user: {
        chatters: {
          dcpesses: {
            id: '473294395',
            login: 'dcpesses',
            display_name: 'dcpesses',
            type: '',
            broadcaster_type: '',
            description: 'Don\'t mind me, I\'m just here for the Jackbox games and to support my peeps. ',
            profile_image_url: 'profile_image-300x300.png',
            offline_image_url: '',
            view_count: 0,
            created_at: '2019-11-18T10:47:34Z'
          }
        },
        whisperStatus: {
          dcpesses: {
            login: 'dcpesses',
            response: {
              msg: 'Error 400 sending to @dcpesses: A user cannot whisper themself',
              status: 400,
              error: {
                error: 'Bad Request',
                status: 400,
                message: 'A user cannot whisper themself',
                player: {
                  id: '473294395',
                  username: 'dcpesses'
                }
              }
            }
          }
        }
      }
    });
    const actionHandler = vi.fn();
    const removeHandler = vi.fn();
    const {container} = render(
      <Provider store={store}>
        <PlayerQueueCard
          btnProps={{
            label: 'Move 2 Interesting',
            onClick: actionHandler,
          }}
          onRemoveUser={removeHandler}
          onSendCode={null}
          queueName={'interested'}
          prioritySeat={false}
          relativeTime={null}
          showSendButton={false}
          username={'dcpesses'}
        />
      </Provider>
    );
    expect(container).toMatchSnapshot();

    const button1 = screen.getByTitle('Remove');
    fireEvent.click(button1);
    expect(removeHandler).toHaveBeenCalled();
    const button2 = screen.getByTitle('Move 2 Interesting');
    fireEvent.click(button2);
    expect(actionHandler).toHaveBeenCalled();
  });

  test('Should render with priority seat and relative time stamps', () => {
    vi.useFakeTimers();
    vi.spyOn(Date, 'now').mockReturnValue(1445470140000); // October 21, 2015 4:29:00 PM PST
    const {container} = render(
      <Provider store={store}>
        <PlayerQueueCard
          prioritySeat={true}
          relativeTime="Now"
        />
      </Provider>
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText('Now')).not.toBeNull();
    vi.useRealTimers();
  });
});

