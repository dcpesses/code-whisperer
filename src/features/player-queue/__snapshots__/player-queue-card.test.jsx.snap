// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayerQueueCard > Should render as a user in the interested queue 1`] = `
<div>
  <div
    class="player-queue-card p-2 mb-0 small lh-1 border-bottom w-100 raleway-font fw-medium border rounded bg-dark-subtle position-relative"
  >
    <div
      class="d-flex justify-content-between"
    >
      <button
        class="btn btn-base border-0 d-flex flex-row"
        title="View Additional User Information"
      >
        <img
          alt="dcpesses"
          class="rounded-circle user-profile-image"
          src="profile_image-300x300.png"
        />
         
        <div
          class="flex-column ms-1 text-start"
        >
          <strong
            class="text-body-emphasis fs-4 saira-condensed fw-bold d-inline-block"
          >
            dcpesses
             
          </strong>
          <span
            class="text-purple-2 d-block smaller fw-semibold"
          />
        </div>
      </button>
      <div
        class="d-flex flex-row"
      >
        <button
          class="btn btn-icon remove-user"
          title="Remove"
        >
          <i
            class="bi-trash3-fill"
          />
        </button>
        <button
          class="btn btn-icon"
          title="Add 2 Playing"
        >
          <i
            class="bi-plus-circle-fill text-purple-4"
          />
        </button>
      </div>
    </div>
    <div
      class="extended-info-pane p-0 m-0 collapse"
      id="player-queue-card-info-dcpesses"
    >
      <div
        class="player-queue-card-info border-top mt-2 pt-2 px-1 lh-base"
      >
        <span
          class="d-block"
        >
          <i
            aria-label="cake icon"
            class="bi-cake2-fill text-purple-2"
          />
           
          November 18, 2019
           
          <span
            class="text-purple-2"
          >
            (
            5 years ago
            )
          </span>
        </span>
        <div
          class="pt-2 fw-semibold lh-sm"
        >
          Don't mind me, I'm just here for the Jackbox games and to support my peeps. 
        </div>
        <div
          class="d-block mx-auto pt-2"
        >
          <a
            class="d-inline-block ttv-link fw-bold"
            href="https://www.twitch.com/dcpesses"
            rel="noreferrer"
            target="_blank"
          >
            View Profile
          </a>
        </div>
      </div>
    </div>
    <div
      class="d-block text-center position-absolute bottom-0 start-50 translate-middle-x"
    >
      <button
        aria-controls="player-queue-card-info-dcpesses"
        aria-expanded="false"
        class="btn badge rounded-pill user-info-toggle"
      >
        View More
         
        <i
          class="bi-caret-down-fill "
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`PlayerQueueCard > Should render as a whispered user in the playing queue 1`] = `
<div>
  <div
    class="player-queue-card p-2 mb-0 small lh-1 border-bottom w-100 raleway-font fw-medium border rounded bg-dark-subtle position-relative"
  >
    <div
      class="d-flex justify-content-between"
    >
      <button
        class="btn btn-base border-0 d-flex flex-row"
        title="View Additional User Information"
      >
        <img
          alt="dcpesses"
          class="rounded-circle user-profile-image"
          src="profile_image-300x300.png"
        />
         
        <div
          class="flex-column ms-1 text-start"
        >
          <strong
            class="text-body-emphasis fs-4 saira-condensed fw-bold d-inline-block"
          >
            dcpesses
             
          </strong>
          <span
            class="text-purple-2 d-block smaller fw-semibold"
          />
        </div>
      </button>
      <div
        class="d-flex flex-row"
      >
        <button
          class="btn btn-icon remove-user"
          title="Remove"
        >
          <i
            class="bi-trash3-fill"
          />
        </button>
        <button
          class="btn btn-icon"
          title="Move 2 Interesting"
        >
          <i
            class="bi-plus-circle-fill text-purple-4"
          />
        </button>
      </div>
    </div>
    <div
      class="extended-info-pane p-0 m-0 collapse"
      id="player-queue-card-info-dcpesses"
    >
      <div
        class="player-queue-card-info border-top mt-2 pt-2 px-1 lh-base"
      >
        <div
          class="fade alert alert-danger show"
          role="alert"
        >
          <i
            class="bi-exclamation-triangle-fill"
          />
           
          Error 400 sending to @dcpesses: A user cannot whisper themself
        </div>
        <span
          class="d-block"
        >
          <i
            aria-label="cake icon"
            class="bi-cake2-fill text-purple-2"
          />
           
          November 18, 2019
           
          <span
            class="text-purple-2"
          >
            (
            5 years ago
            )
          </span>
        </span>
        <div
          class="pt-2 fw-semibold lh-sm"
        >
          Don't mind me, I'm just here for the Jackbox games and to support my peeps. 
        </div>
        <div
          class="d-block mx-auto pt-2"
        >
          <a
            class="d-inline-block ttv-link fw-bold"
            href="https://www.twitch.com/dcpesses"
            rel="noreferrer"
            target="_blank"
          >
            View Profile
          </a>
        </div>
      </div>
    </div>
    <div
      class="d-block text-center position-absolute bottom-0 start-50 translate-middle-x"
    >
      <button
        aria-controls="player-queue-card-info-dcpesses"
        aria-expanded="false"
        class="btn badge rounded-pill user-info-toggle"
      >
        View More
         
        <i
          class="bi-caret-down-fill "
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`PlayerQueueCard > Should render with priority seat and relative time stamps 1`] = `
<div>
  <div
    class="player-queue-card p-2 mb-0 small lh-1 border-bottom w-100 raleway-font fw-medium border rounded bg-dark-subtle position-relative"
  >
    <div
      class="d-flex justify-content-between"
    >
      <button
        class="btn btn-base border-0 d-flex flex-row"
        title="View Additional User Information"
      >
        <i
          class="bi-person-fill rounded-circle user-profile-image bg-secondary bg-gradient text-white-50"
        />
         
        <div
          class="flex-column ms-1 text-start"
        >
          <strong
            class="text-warning-emphasis fs-4 saira-condensed fw-bold d-inline-block"
          >
             
            <span
              class="align-self-center fs-6"
              title="Priority seat redemption"
            >
              <i
                class="bi-stars"
              />
            </span>
          </strong>
          <span
            class="text-purple-2 d-block smaller fw-semibold"
          >
            Now
          </span>
        </div>
      </button>
      <div
        class="d-flex flex-row"
      >
        <button
          class="btn btn-icon remove-user"
          title="Remove"
        >
          <i
            class="bi-trash3-fill"
          />
        </button>
        <button
          class="btn btn-icon"
          title=""
        />
      </div>
    </div>
    <div
      class="extended-info-pane p-0 m-0 collapse"
      id="player-queue-card-info-null"
    >
      <div
        class="player-queue-card-info border-top mt-2 pt-2 px-1 lh-base"
      >
        <span
          class="d-block"
        >
          <i
            aria-label="cake icon"
            class="bi-cake2-fill text-purple-2"
          />
           
           
          <span
            class="text-purple-2"
          >
            (
            )
          </span>
        </span>
        <div
          class="pt-2 fw-semibold lh-sm"
        />
        <div
          class="d-block mx-auto pt-2"
        >
          <a
            class="d-inline-block ttv-link fw-bold"
            href="https://www.twitch.com/null"
            rel="noreferrer"
            target="_blank"
          >
            View Profile
          </a>
        </div>
      </div>
    </div>
    <div
      class="d-block text-center position-absolute bottom-0 start-50 translate-middle-x"
    >
      <button
        aria-controls="player-queue-card-info-null"
        aria-expanded="false"
        class="btn badge rounded-pill user-info-toggle"
      >
        View More
         
        <i
          class="bi-caret-down-fill "
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`PlayerQueueCard > Should render without error 1`] = `
<div>
  <div
    class="player-queue-card p-2 mb-0 small lh-1 border-bottom w-100 raleway-font fw-medium border rounded bg-dark-subtle position-relative"
  >
    <div
      class="d-flex justify-content-between"
    >
      <button
        class="btn btn-base border-0 d-flex flex-row"
        title="View Additional User Information"
      >
        <i
          class="bi-person-fill rounded-circle user-profile-image bg-secondary bg-gradient text-white-50"
        />
         
        <div
          class="flex-column ms-1 text-start"
        >
          <strong
            class="text-body-emphasis fs-4 saira-condensed fw-bold d-inline-block"
          >
             
          </strong>
          <span
            class="text-purple-2 d-block smaller fw-semibold"
          />
        </div>
      </button>
      <div
        class="d-flex flex-row"
      >
        <button
          class="btn btn-icon remove-user"
          title="Remove"
        >
          <i
            class="bi-trash3-fill"
          />
        </button>
        <button
          class="btn btn-icon"
          title=""
        />
      </div>
    </div>
    <div
      class="extended-info-pane p-0 m-0 collapse"
      id="player-queue-card-info-null"
    >
      <div
        class="player-queue-card-info border-top mt-2 pt-2 px-1 lh-base"
      >
        <span
          class="d-block"
        >
          <i
            aria-label="cake icon"
            class="bi-cake2-fill text-purple-2"
          />
           
           
          <span
            class="text-purple-2"
          >
            (
            )
          </span>
        </span>
        <div
          class="pt-2 fw-semibold lh-sm"
        />
        <div
          class="d-block mx-auto pt-2"
        >
          <a
            class="d-inline-block ttv-link fw-bold"
            href="https://www.twitch.com/null"
            rel="noreferrer"
            target="_blank"
          >
            View Profile
          </a>
        </div>
      </div>
    </div>
    <div
      class="d-block text-center position-absolute bottom-0 start-50 translate-middle-x"
    >
      <button
        aria-controls="player-queue-card-info-null"
        aria-expanded="false"
        class="btn badge rounded-pill user-info-toggle"
      >
        View More
         
        <i
          class="bi-caret-down-fill "
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`PlayerQueueCard > Should render without error 2`] = `
<div>
  <div
    class="player-queue-card p-2 mb-0 small lh-1 border-bottom w-100 raleway-font fw-medium border rounded bg-dark-subtle position-relative"
  >
    <div
      class="d-flex justify-content-between"
    >
      <button
        class="btn btn-base border-0 d-flex flex-row"
        title="View Additional User Information"
      >
        <i
          class="bi-person-fill rounded-circle user-profile-image bg-secondary bg-gradient text-white-50"
        />
         
        <div
          class="flex-column ms-1 text-start"
        >
          <strong
            class="text-body-emphasis fs-4 saira-condensed fw-bold d-inline-block"
          >
             
          </strong>
          <span
            class="text-purple-2 d-block smaller fw-semibold"
          />
        </div>
      </button>
      <div
        class="d-flex flex-row"
      >
        <button
          class="btn btn-icon remove-user"
          title="Remove"
        >
          <i
            class="bi-trash3-fill"
          />
        </button>
        <button
          class="btn btn-icon"
          title=""
        />
      </div>
    </div>
    <div
      class="extended-info-pane p-0 m-0 collapsing"
      id="player-queue-card-info-null"
      style="height: 0px;"
    >
      <div
        class="player-queue-card-info border-top mt-2 pt-2 px-1 lh-base"
      >
        <span
          class="d-block"
        >
          <i
            aria-label="cake icon"
            class="bi-cake2-fill text-purple-2"
          />
           
           
          <span
            class="text-purple-2"
          >
            (
            )
          </span>
        </span>
        <div
          class="pt-2 fw-semibold lh-sm"
        />
        <div
          class="d-block mx-auto pt-2"
        >
          <a
            class="d-inline-block ttv-link fw-bold"
            href="https://www.twitch.com/null"
            rel="noreferrer"
            target="_blank"
          >
            View Profile
          </a>
        </div>
      </div>
    </div>
    <div
      class="d-block text-center position-absolute bottom-0 start-50 translate-middle-x"
    >
      <button
        aria-controls="player-queue-card-info-null"
        aria-expanded="true"
        class="btn badge rounded-pill user-info-toggle"
      >
        View Less
         
        <i
          class="bi-caret-up-fill "
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`PlayerQueueCard > Should render without error 3`] = `
<div>
  <div
    class="player-queue-card p-2 mb-0 small lh-1 border-bottom w-100 raleway-font fw-medium border rounded bg-dark-subtle position-relative"
  >
    <div
      class="d-flex justify-content-between"
    >
      <button
        class="btn btn-base border-0 d-flex flex-row"
        title="View Additional User Information"
      >
        <i
          class="bi-person-fill rounded-circle user-profile-image bg-secondary bg-gradient text-white-50"
        />
         
        <div
          class="flex-column ms-1 text-start"
        >
          <strong
            class="text-body-emphasis fs-4 saira-condensed fw-bold d-inline-block"
          >
             
          </strong>
          <span
            class="text-purple-2 d-block smaller fw-semibold"
          />
        </div>
      </button>
      <div
        class="d-flex flex-row"
      >
        <button
          class="btn btn-icon remove-user"
          title="Remove"
        >
          <i
            class="bi-trash3-fill"
          />
        </button>
        <button
          class="btn btn-icon"
          title=""
        />
      </div>
    </div>
    <div
      class="extended-info-pane p-0 m-0 collapsing"
      id="player-queue-card-info-null"
      style="height: 0px;"
    >
      <div
        class="player-queue-card-info border-top mt-2 pt-2 px-1 lh-base"
      >
        <span
          class="d-block"
        >
          <i
            aria-label="cake icon"
            class="bi-cake2-fill text-purple-2"
          />
           
           
          <span
            class="text-purple-2"
          >
            (
            )
          </span>
        </span>
        <div
          class="pt-2 fw-semibold lh-sm"
        />
        <div
          class="d-block mx-auto pt-2"
        >
          <a
            class="d-inline-block ttv-link fw-bold"
            href="https://www.twitch.com/null"
            rel="noreferrer"
            target="_blank"
          >
            View Profile
          </a>
        </div>
      </div>
    </div>
    <div
      class="d-block text-center position-absolute bottom-0 start-50 translate-middle-x"
    >
      <button
        aria-controls="player-queue-card-info-null"
        aria-expanded="false"
        class="btn badge rounded-pill user-info-toggle"
      >
        View More
         
        <i
          class="bi-caret-down-fill "
        />
      </button>
    </div>
  </div>
</div>
`;
